onnxruntime/core/mickey/blk_q4/prepack_sm80.h

```diff
@@ -0,0 +1,317 @@
+ * @tparam block_size     number of elemenets quantized together
+ * @tparam qbits          number of bits in each quantized element
+ * @tparam Columnwise     true:  elements in a block come from one single column
+ *                        false: elements in a block come from one single row
+ */
+template <
+    typename ElementT,
+    int block_size,
+    int qbits,
+    bool Columnwise,
+    bool ExtraBoundsCheck = false>
+struct BlockwiseQuantization {
+  static_assert(qbits == 4, "Only 4b block quantization is supported!");
+  static_assert(sizeof(ElementT) == 2, "Only 16b floating point types are supported!");
+
+  using QuantBlocking =
+      std::conditional_t<Columnwise,
+                         MatrixShape<block_size, 1>,
+                         MatrixShape<1, block_size>>;
+
+  using ElementW = uint8_t;  // <- Weight is int4, uint8 for two of them
+  // We pack 4 weights into one 16b element, so we can leverage cutlass tile iterators
+  // for async share memory loading, and minimizing bank conflict during matrix loading
+  using ElementWPack = ElementT;
+  using LayoutWPack = ColumnMajorLayout;  // <- layout of packed weight, must be column major
+
+  // Current Ampere kernel use 8b zero point, need to shrink it to 4b in the future
+  using ElementQOffset = uint8_t;
+
+  // Layout of the quantization parameters (scales and zero points)
+  // Major on the dimension that has the most parameters per squarish weight block.
+  // E.g. for column-wise quantization, a [64, 64] block has [2, 64] parameters,
+  // where each row has more data, so we use row major layout so that warp threads
+  // can use less load instructions to load more parameters.
+  using LayoutQmeta =
+      typename std::conditional<Columnwise,
+                                RowMajorLayout, ColumnMajorLayout>::type;
+
+  /**
+   * @brief  Get quantized weight tensor dimensions.
+   * Actual weight type is int4, we use ElementW = uint8 to avoid possible compilation
+   * troubles. Since the layout is column major, we are packing 2 weights in a column
+   * into one int8
+   */
+  static inline auto get_quant_weights_shape(int rows, int columns) {
+    return make_Position((rows + 1) / 2, columns);
+  }
+
+  static inline auto get_quant_meta_shape(int rows, int columns) {
+    return make_Position(rows / QuantBlocking::kRow, columns / QuantBlocking::kColumn);
+  }
+
+  /**
+   * @brief Prepack weight matrix to facilitate matrix loading, depending on MMA
+   * instruction layout.
+   *
+   * The weight matrix is int4, yet we want to leverage existing fp16/bf16
+   * tile loading and MMA layout code in CUTLASS. So we group 4 int4 into 2
+   * bytes, pretending it's fp16. This grouping must be done in a way to be
+   * easily unpacked into tiles that match the MMA instruction layout.
+   * For MMA instruction <16, 8, 16>, each instruction processes 2 8x8 tiles,
+   * vertically stacked on the K dimension. And MmaTensorOpMultiplicandTileIterator
```

Comment:
>K [](http://example.com/codeflow?start=31&length=1)

on M dimension?

Reply:
no, K dimension

Reply:
For the number in "MMA instruction <16, 8, 16>, each instruction processes 2 8x8 tiles, vertically stacked on the K dimension", <16,8,16> maps to M, N, K, and 2 8x8 tiles are the accumulator tile, right?

Reply:
i see, the 2 8x8 tiles is for matrix b.

Reply:
yes! B is 16x8, it's two 8x8 tiles one on top of the other